### PR TITLE
LR-178. Fixed bug with access.

### PR DESCRIPTION
**Pull request by anatoliiibrahimov**
http://ssu-jira.softserveinc.com/browse/LR-178

### Author.

- [x] I have added link to this PR to the JIRA
- [x] I have logged time in JIRA.

### Reviewers.

- [x] @OKrain 
- [x] @brouhaha88 
- [x] @vovka

### Issue summary.

*Not logged in user has access to resource*

### Fix summary. 

Fixed bug by adding before action filter.

### Testing approach.
Included Devise test helpers.
Added two tests to check:
1. Not logged in users don't have access to inner pages.
2. Not logged in users have access to home page.

### CODE REVIEW

**Comment by OKrain** [comment]
maybe u should try <code>bundle exec</code> in CI settings?

File: config/application.rb [hunk 1]
```diff
@@ -1,7 +1,7 @@
 require File.expand_path('../boot', __FILE__)
 
 require 'rails/all'
-
+require 'devise'
```
Comment (vovka) [inline config/application.rb]: Do we need this change?

File: spec/spec_helper.rb [hunk 2]
```diff
@@ -1,4 +1,7 @@
 require 'coveralls'
+require 'rspec/rails'
```
Comment (vovka) [inline spec/spec_helper.rb]: Do we need this change?

**Comment by vovka** [comment]
Please, add some test to check that:
1. Not logged in users don't have access to inner pages.
2. Not logged in users have access to home page.

**Comment by vovka** [comment]
What about other test which existed before your change? Don't they fail?

File: config/application.rb [hunk 3]
```diff
@@ -22,5 +22,6 @@ class Application < Rails::Application
 
     # Do not swallow errors in after_commit/after_rollback callbacks.
     config.active_record.raise_in_transactional_callbacks = true
+    if defined?(Bundler)
```
Comment (vovka) [inline config/application.rb]: Do we need this change?

Comment (superNazo) [inline config/application.rb]: u've missed one end for if statement, build is red for this reason

File: spec/controllers/quizzes_controller_spec.rb [hunk 4]
```diff
@@ -54,6 +54,7 @@
 
   describe "GET #new" do
     it "assigns a new quiz as @quiz" do
+      quiz = Quiz.create! valid_attributes
```
Comment (vovka) [inline spec/controllers/quizzes_controller_spec.rb]: Codacy Issue found: [Useless assignment to variable - `quiz`.](https://www.codacy.com/app/vovka/video_quiz/file/2360561850/issues/source?bid=3205144&fileBranchId=3218191#l57)

File: spec/requests/quizzes_spec.rb [hunk 5]
```diff
@@ -4,7 +4,7 @@
   describe "GET /quizzes" do
     it "works! (now write some real specs)" do
       get quizzes_path
-      expect(response).to have_http_status(200)
+      expect(response).to have_http_status(302)
```
Comment (vovka) [inline spec/requests/quizzes_spec.rb]: This test is an example and should be removed. You have to write your own tests to check that logged in user gets access to quizzes page. Plus my previous comment about writing tests.

**Comment by OKrain** [comment]
+1

File: spec/controllers/quizzes_controller_spec.rb [hunk 6]
```diff
@@ -54,6 +71,7 @@
 
   describe "GET #new" do
     it "assigns a new quiz as @quiz" do
+      Quiz.create! valid_attributes
```
Comment (vovka) [inline spec/controllers/quizzes_controller_spec.rb]: Why did you add this line?

File: test/test_helper.rb [hunk 7]
```diff
@@ -6,6 +6,7 @@
 require 'rails/test_help'
 
 class ActiveSupport::TestCase
+  include Devise::TestHelpers
```
Comment (vovka) [inline test/test_helper.rb]: We don't use test in `/test` folder, do we?

File: spec/controllers/quizzes_controller_spec.rb [hunk 8]
```diff
@@ -39,9 +39,26 @@
...
+      expect(response.status).to eq 302
+    end
+
+    it "if user sign in show quizzes" do
+      user = User.create!(email: "user@example.org", name: "Name")
+      sign_in user
+
+      get :index, {}, valid_session
+
+      expect(response.status).to eq 200
```
Comment (vovka) [inline spec/controllers/quizzes_controller_spec.rb]: It's true

File: spec/controllers/quizzes_controller_spec.rb [hunk 9]
```diff
@@ -54,7 +71,7 @@
 
   describe "GET #new" do
     it "assigns a new quiz as @quiz" do
-      get :new, {}, valid_session
+      get :new, {quiz: valid_attributes}, valid_session
```
Comment (vovka) [inline spec/controllers/quizzes_controller_spec.rb]: This change just makes test pending:
3) QuizzesController GET #new assigns a new quiz as @quiz
# Add a hash of attributes valid for your model
     # ./spec/controllers/quizzes_controller_spec.rb:73

**Comment by brouhaha88** [comment]
I Think its ready

File: spec/controllers/quizzes_controller_spec.rb [hunk 10]
```diff
@@ -54,7 +71,8 @@
 
   describe "GET #new" do
     it "assigns a new quiz as @quiz" do
-      get :new, {}, valid_session
+      quiz = Quiz.new! valid_attributes
```
Comment (vovka) [inline spec/controllers/quizzes_controller_spec.rb]: Good try. But no:) Tests is still pending.

File: spec/controllers/quizzes_controller_spec.rb [hunk 11]
```diff
@@ -39,9 +39,26 @@
   describe "GET #index" do
     it "assigns all quizzes as @quizzes" do
       quiz = Quiz.create! valid_attributes
+
```
Comment (YUzhva) [inline spec/controllers/quizzes_controller_spec.rb]: Чому...

File: spec/controllers/quizzes_controller_spec.rb [hunk 12]
```diff
@@ -39,9 +39,26 @@
...
       get :index, {}, valid_session
+
```
Comment (YUzhva) [inline spec/controllers/quizzes_controller_spec.rb]: в тестах...

File: spec/controllers/quizzes_controller_spec.rb [hunk 13]
```diff
@@ -39,9 +39,26 @@
...
       expect(assigns(:quizzes)).to eq([quiz])
     end
+
+    it "redirect to home page if not sign in" do
+      get :index, {}, valid_session
+
```
Comment (YUzhva) [inline spec/controllers/quizzes_controller_spec.rb]: три пусті рядки?

File: spec/controllers/quizzes_controller_spec.rb (hunk 11 above)
Comment (YUzhva) [inline spec/controllers/quizzes_controller_spec.rb]: 1 =>

File: spec/controllers/quizzes_controller_spec.rb (hunk 12 above)
Comment (YUzhva) [inline spec/controllers/quizzes_controller_spec.rb]: 2 =>

File: spec/controllers/quizzes_controller_spec.rb (hunk 13 above)
Comment (YUzhva) [inline spec/controllers/quizzes_controller_spec.rb]: 3 =>

File: spec/controllers/quizzes_controller_spec.rb [hunk 14]
```diff
@@ -39,9 +39,26 @@
...
+      expect(response.status).to eq 302
+    end
+
+    it "if user sign in show quizzes" do
+      user = User.create!(email: "user@example.org", name: "Name")
+      sign_in user
+
```
Comment (YUzhva) [inline spec/controllers/quizzes_controller_spec.rb]: 4 =>

File: spec/controllers/quizzes_controller_spec.rb [hunk 15]
```diff
@@ -39,9 +39,26 @@
...
+      get :index, {}, valid_session
+
```
Comment (YUzhva) [inline spec/controllers/quizzes_controller_spec.rb]: 5ть пустих рядочків хочуть потрапити в продакшн.
https://github.com/superNazo/video_quiz/blob/bugfix/restrict_access/spec/controllers/quizzes_controller_spec.rb

Comment (anatoliiibrahimov) [inline spec/controllers/quizzes_controller_spec.rb]: це просто відокремлення всіх фаз тесту.
